aips/account_abstraction.md

```diff
@@ -0,0 +1,204 @@
+discussions-to (*optional): <a url pointing to the official discussion thread>
+Status: Draft
+last-call-end-date (*optional): <mm/dd/yyyy the last date to leave feedbacks and reviews>
+type: Framework
+created: 10/01/2024
+updated (*optional): <mm/dd/yyyy>
+requires (*optional): <AIP number(s)>
+---
+
+# AIP-X - Account Abstraction
+
+## Summary
+
+Account Abstraction (AA) on Aptos allows any account to be authenticated through move code in addition to existing 
+authentication schemes supported by native code. This AIP will introduce a new authenticator called `Abstraction` that
+delegates the authentication verification to AptosVM running move-based authentication logic.
+
+### Out of scope
+
+How to properly design an authentication function/module is out of scope.
+The `PayMaster`, aka, online gas payer, is not included in the design of AA on Aptos.
+
+## High-level Overview
+The core concept is to enable the move function to authenticate account signers, moving beyond the limitations of native Rust authenticators, which only support a restricted set of cryptographic schemes. To accomplish this, we propose transferring the authentication process from native Rust code to the AptosVM layer for each abstracted account. The solution we propose involves the following steps:
+
+1. Store the `FunctionInfo` of the customized Move authentication function in a new resource at the AA (Abstract Account) address.
+2. When the account uses an AbstractionAuthenticator, the Move function stored in step 1 will be executed (leveraging native dynamic dispatch) to perform the authentication check.
+
+### Why is this approach superior to other options?
+
+- No major security concerns: Users who opt for AA are responsible for managing their account’s security.
+- Scalable: This approach can easily scale as it decouples authentication from rigid native cryptographic checks and supports multiple authentication functions.
+- Audit-proof and safe: The dispatchable function mechanism has already been audited and is widely used in the FA standard, ensuring its security.
+- Rich transaction context: The Transaction Context already provides comprehensive information about the current transaction, which can be leveraged for robust authentication.
+
+This solution enhances flexibility and scalability while maintaining a secure and efficient authentication process.
+
+## Impact
+
+Account Abstraction(AA) aims to support dispatchable authentication **at the level of smart contract.**
+The account can set up in such a way that the transaction can be authorized with complicated move smart contract instead of native account authenticators provided by Aptos blockchain core, enabling a series of possibilities:
+
+- Multisig v2 account
+- Sandbox or temporary account to isolate assets/risks, etc.
+- Restricted permissions granted to trusted applications so users don’t need to sign every transaction.
+- Delegation of asset management to a different account (semi-custodial model)
```

Comment:
it'd be nice if a bit more detail can be provided here for each item. i'm curious how this can be related to multisig v2 accounts

Reply:
multisig v2 uses `MultisigAccount` as the account resource. 
And the sender of multisig v2 transaction is a normal sender. They could be an AA account. I think basically if you want to migrate an account from multisig v2 to AA account. We have to find a way first to convert multisig account to a normal account with AA auth to seamlessly migrate. I think you are asking about this.

But this^ seems unrelated to AA design itself. It would be worth another AIP to explain this migration option.

Reply:
Enabling AA for an account means registering an account resource that contains a dispatchable auth function, right?

If enabling AA can be done via say a private entry function call, I think there won't be any special treatment needed for multisig v2 accounts. Since multisig v2 accounts are just accounts that happen to have a MulsitigAccount resource, they can just call the AA enabling function like any other accounts.

Reply:
I think they can coexist. Any problem for that. Once you get the master signer in any means, you can opt to remove any auth resources.